presto-main/src/main/java/com/facebook/presto/execution/QueryLimitMemoryRevokingScheduler.java

```diff
@@ -0,0 +1,157 @@
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.PriorityQueue;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Function;
+
+import static com.facebook.presto.execution.MemoryRevokingUtils.getMemoryPools;
+import static java.util.Objects.requireNonNull;
+
+public class QueryLimitMemoryRevokingScheduler
+{
+    private static final Logger log = Logger.get(QueryLimitMemoryRevokingScheduler.class);
+
+    private final ScheduledExecutorService taskManagementExecutor;
+    private final Function<QueryId, QueryContext> queryContextSupplier;
+
+    private final AtomicBoolean checkPending = new AtomicBoolean();
+    private final List<MemoryPool> memoryPools;
+    private final MemoryPoolListener memoryPoolListener = this::onMemoryReserved;
+
+    @Inject
+    public QueryLimitMemoryRevokingScheduler(
+            LocalMemoryManager localMemoryManager,
+            SqlTaskManager sqlTaskManager,
+            TaskManagementExecutor taskManagementExecutor)
+    {
+        this(
+                ImmutableList.copyOf(getMemoryPools(localMemoryManager)),
+                requireNonNull(sqlTaskManager, "sqlTaskManager cannot be null")::getQueryContext,
+                requireNonNull(taskManagementExecutor, "taskManagementExecutor cannot be null").getExecutor());
+        log.debug("Using QueryLimitMemoryRevokingScheduler spilling strategy");
+    }
+
+    @VisibleForTesting
+    QueryLimitMemoryRevokingScheduler(
+            List<MemoryPool> memoryPools,
+            Function<QueryId, QueryContext> queryContextSupplier,
+            ScheduledExecutorService taskManagementExecutor)
+    {
+        this.memoryPools = ImmutableList.copyOf(requireNonNull(memoryPools, "memoryPools is null"));
+        this.queryContextSupplier = requireNonNull(queryContextSupplier, "queryContextSupplier is null");
+        this.taskManagementExecutor = requireNonNull(taskManagementExecutor, "taskManagementExecutor is null");
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        registerPoolListeners();
+    }
+
+    @PreDestroy
+    public void stop()
+    {
+        memoryPools.forEach(memoryPool -> memoryPool.removeListener(memoryPoolListener));
+    }
+
+    @VisibleForTesting
+    void registerPoolListeners()
+    {
+        memoryPools.forEach(memoryPool -> memoryPool.addListener(memoryPoolListener));
+    }
+
+    private void onMemoryReserved(MemoryPool memoryPool, QueryId queryId, long queryTotalMemoryUse)
+    {
+        try {
+            QueryContext queryContext = queryContextSupplier.apply(queryId);
+            long maxTotalMemory = queryContext.getMaxTotalMemory();
+            if (queryTotalMemoryUse <= maxTotalMemory) {
+                return;
+            }
+
+            if (checkPending.compareAndSet(false, true)) {
```

Comment:
Seems like this is only going to allow one query to be revoked at a time and would permit queries to skip revoking if another unrelated query is being revoked. One option I suppose would be to use a concurrent set of query Id's being revoked, but another might be to enforce this whole thing inside of the `QueryContext` itself since you could trigger that reliably while holding the lock.

Reply:
That's a great catch.  To clarify your suggetion, do you mean checkPending should go in `QueryContext` or the whole `revokeMemory()` method?